language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/semantictokens/SemanticTokensVisitor.java

```diff
@@ -650,24 +654,13 @@ private void addSemanticToken(Node node, int type, int modifiers, boolean proces
      * @param modifiers    Semantic token type modifiers' index
      */
     private void handleReferences(LinePosition linePosition, int length, int type, int modifiers) {
-        Optional<SemanticModel> semanticModel = this.semanticTokensContext.currentSemanticModel();
-        if (semanticModel.isEmpty()) {
-            return;
-        }
-
-        Optional<Document> docOptional = this.semanticTokensContext.currentDocument();
-        if (docOptional.isEmpty()) {
-            return;
-        }
-
-        Optional<String> path =
-                this.semanticTokensContext.workspace().relativePath(this.semanticTokensContext.filePath());
-        if (path.isEmpty()) {
+        if (semanticModel == null || path == null) {
```

Comment:
Avoid null check and use Optional checks as best practice. Please refer to the previous comments regarding this. You don't need to cache the path since we refer to the current path and that will be able to access via the context

Reply:
> Avoid null check and use Optional checks as best practice. Please refer to the previous comments regarding this. You don't need to cache the path since we refer to the current path and that will be able to access via the context

We have been using the context to access the filePath. According to the performance analysis, `this.semanticTokensContext.workspace().relativePath()` call takes a considerable time when we have repetitive calls. Hence, cached the value at the visitor level.